File: WallBreakers/SimpleBitManipulation/NumComplement.py
class Solution:
  def findComplement(self, num: int) -> int:
    result = ''
    binStr = bin(num)
    for b in binStr[2:]:
      if b == '0': 
        result += '1'
      else: 
        result += '0'
    return int(result,2)
  # def findComplement(self, num: int) -> int:
  #   size = len(bin(num))-2 # exclude '0b'
  #   # return num ^ (2**num.bit_length()-1)
  # return num ^ (2**size - 1)

File: WallBreakers/SimpleBitManipulation/NumComplement.java
class Solution {
  // public int findComplement(int num) {
  //   StringBuilder sb = new StringBuilder();
  //   char[] binStr = (Integer.toBinaryString(num)).toCharArray();
  //   for (char c : binStr) {
  //     if (c == '0') {
  //       sb.append("1");
  //     } else {
  //       sb.append("0");
  //     }
  //   }
  //   return Integer.parseInt(sb.toString(), 2);
  // }
  public int findComplement(int num) {
    // invert bin then mask with num of bits
    // return ~num & ((Integer.highestOneBit(num) << 1) - 1);
    return num ^ (int)(Math.pow(2, (int)(Math.log(num) / Math.log(2) + 1))-1);
  }
}